### PR TITLE
Jetpack Assistant: Add UI

### PR DESCRIPTION
**Pull request by robertf4**
#### Changes proposed in this Pull Request:

This PR adds the Jetpack Assistant. It is the content of [the original PR](https://github.com/Automattic/jetpack/pull/17691) rebased against the repo reorg in master and replaces that PR. 

Site type prompt:
image
image

WooCommerce prompt:
image
image

Monitor prompt:
image
image

Related posts prompt:
image
image

Creative Mail prompt:
image
image

Site accelerator prompt:
image
image

Summary screen with no recommendations enabled/installed:
image
image

Summary screen with all recommendations enabled/installed:
image
image

Backup Daily sidebar upsell:
image

Fallback sidebar upsell (shown if disabled in wpcom or wpcom can't be reached):
image

Enable one-click restores sidebar prompt:
image

Manage your security sidebar prompt:
image

Apps sidebar recommendation:
image

#### Jetpack product discussion

MT: p1HpG7-asW-p2
Design: p6TEKc-46C-p2

#### Testing instructions:

Initial prompts flow:
1. Start with a free site and navigate to `/wp-admin/admin.php?page=jetpack#/recommendations`.
2. Verify that you are redirected to `/wp-admin/admin.php?page=jetpack#/recommendations/site-type`.
3. Ensure that you have not selected "Store" and click "Continue".
4. Verify that you navigate to `#/recommendations/monitor`.
5. Click back in your browser and select "Store", then click "Continue".
6. Verify that you navigate to `#/recommendations/woocommerce`.
7. Navigate to `/wp-admin/index.php` and then back to `/wp-admin/admin.php?page=jetpack#/recommendations`.
8. Verify that you are redirected to `#/recommendations/woocommerce`.
9. Choose "Decide later for each of the prompts and verify that you are taken successively through `#/recommendations/monitor`, `#/recommendations/related-posts`, `#/recommendations/creative-mail`, and `#/recommendations/site-accelerator`.
10. Verify that the summary screen has no recommendations enabled.
11. Now return to `#/recommendations/monitor` and go back through the flow, selecting some recommendations and skipping others. Verify that the summary screen reflects your choices.
12. Enable the remaining recommendations on the summary screen and verify that it updates to reflect these changes.
13. Check the various "Learn more" and "Settings" buttons for each feature and make sure they are appropriate for that feature.
14. Check that the "View all Jetpack features' link works.

Summary upsell flows:
1. On a free site visit `/wp-admin/admin.php?page=jetpack#/recommendations/summary`.
2. Verify that an upsell prompt for Backup Daily shows.
3. Sandbox your site and in wpcom edit `/wp-content/rest-api-plugins/endpoints/jetpack-recommendations.php` so that the `WP_REST_Response` in `get_upsell()` returns `'hide_upsell' => true`.
4. Reload `#/recommendations/summary` and verify that the fallback sidebar upsell is displayed (no prices or product names should show).
5. Upgrade the site to Backup Daily.
6. Return to `#/recommendations/summary`. If the app sidebar prompt is displaying, wait a few minutes for Rewind to update, then reload the page.
7. Verify that the one-click restores prompt is displaying.
8. Use the prompt to add your credentials and then return to `#/recommendations/summary`. Verify that the "Manage your security" prompt displays in the sidebar and that the link works.
9. Start another free site and upgrade it to a paid plan that doesn't include backup or scan (e.g. anti-spam).
10. Navigate back to `#/recommendations/summary` and verify that the app sidebar prompt shows and that the app badges work.

Jetpack plugin upgrade flow:
1. Start with a site on an old version of Jetpack that includes the Setup Wizard (9.3 meets this criteria).
2. Use the Code Snippets plugin and add the following snippet to enable the Setup Wizard:
```
add_filter( 'jetpack_show_setup_wizard', '__return_true' );
add_filter( 'jetpack_pre_connection_prompt_helpers', '__return_true' );
```
3. Navigate to `/wp-admin/admin.php?page=jetpack#/setup/features` and click "I'm done for now" to complete the Setup Wizard.
4. On that site upgrade Jetpack so that it includes the Recommendations.
5. Verify that the Recommendations are not available from the dashboard, and that `#/recommendations` redirects to `#/dashboard`
6. Perform the above steps again but this time do not complete the Setup Wizard and verify that the Recommendations are available.

#### Proposed changelog entry for your changes:

Recommendations: add this new wizard to help you get started with recommended features of Jetpack.

### CODE REVIEW

File: projects/plugins/jetpack/_inc/client/recommendations/sidebar/mobile-app/index.jsx
Comment (jeherve) [inline projects/plugins/jetpack/_inc/client/recommendations/sidebar/mobile-app/index.jsx:25]: Should it be "dashboard"?

```suggestion
		__( 'Manage multiple sites from one dashboard', 'jetpack' ),
```

File: projects/plugins/jetpack/_inc/client/recommendations/sidebar/mobile-app/index.jsx
Comment (jeherve) [inline projects/plugins/jetpack/_inc/client/recommendations/sidebar/mobile-app/index.jsx:28]: Should we include "Refined post and page editor" here, now that the editor in wpcom is an iFramed version of the wp-admin editor? Same with "Free stock photo library", now that the feature is available in wp-admin?

File: projects/plugins/jetpack/_inc/client/recommendations/sidebar/one-click-restores/index.jsx
Comment (jeherve) [inline projects/plugins/jetpack/_inc/client/recommendations/sidebar/one-click-restores/index.jsx:54]: Should we say "you", since that's something you do on your own?

```suggestion
								'Get the most out of your %s. One-click restores ensure you’ll be able to easily restore your site, if anything goes wrong.',
```

**Review by jeherve — CHANGES_REQUESTED** [review]
I gave this a try, and would have some questions.

********

The Progress status for the install buttons seems off a bit:

image

*********

When I tried to install Creative Contact, the installation failed silently (the button went back to "Install"), and here is the response I had in my console:

`Unable to install creative-mail-by-constant-contact: folder_exists`

The plugin had indeed been installed, but the installation had failed. It would be nice to surface any errors in the UI.

**********

When I tried to install Creative Contact again, the installation worked, but there was no indication of it in the UI. The line with the prompt to install disappeared, but that was the only hint that something had happened. I think it would be nice to show a notice with a link to the plugin wizard maybe?

Update: I see now that a line was added to the "enabled" section. I had missed that, I didn't expect that.

**********

I think the app store buttons could do with some extra padding (this was in Chrome)

image

**********

Maybe we should stay focused on "Dashboard" when in recommendations?

image

File: projects/plugins/jetpack/_inc/client/recommendations/feature-summary/index.jsx
Comment (robertf4) [reply projects/plugins/jetpack/_inc/client/recommendations/feature-summary/index.jsx:93]: That is intentional as it matches the Figma. `ExternalLink` only displays the icon when you set `icon=true`.

File: projects/plugins/jetpack/_inc/client/components/number-format/index.js
Comment (robertf4) [reply projects/plugins/jetpack/_inc/client/components/number-format/index.js:36]: Good idea, I moved it to a `lib` folder.

File: projects/plugins/jetpack/package.json
Comment (robertf4) [reply projects/plugins/jetpack/package.json:72]: I am not sure how it ended up in here, but I don't think it is needed. I have removed it.

File: projects/plugins/jetpack/package.json
Comment (robertf4) [reply projects/plugins/jetpack/package.json:70]: I don't think so. I updated this because `@automattic/components` depends on having >=2.0. I can't find any usage of the package in Jetpack, and the changes were all renaming or removing variable names so I think the SCSS compiler would be failing if any of them was being used.

So the only change possible would be a change in a color somewhere, but looking around the plugin I have not noticed anything obvious.

File: projects/plugins/jetpack/_inc/client/state/products/reducer.js
Comment (robertf4) [reply projects/plugins/jetpack/_inc/client/state/products/reducer.js:63]: At one point I was getting a null exception from this when it was accessed before it was initialized, but it's actually not being used anymore since I made a dedicated endpoint to get the upsell info. So I've reverted this change.

File: projects/plugins/jetpack/_inc/client/recommendations/sidebar/mobile-app/apps-badge.jsx
Comment (robertf4) [reply projects/plugins/jetpack/_inc/client/recommendations/sidebar/mobile-app/apps-badge.jsx]: Here I've just taken the component from Calypso and modified it slightly to work within Jetpack, so this is just me taking the behavior there.

In the event that loading the badge fails this component will default to using the local badge. For both iOS and Android the locale is in the URL. So if either the URL changes or if the language is not supported it will end up using the default English ones that are available locally.

File: projects/plugins/jetpack/_inc/client/recommendations/sidebar/mobile-app/index.jsx
Comment (jeffgolenski) [reply projects/plugins/jetpack/_inc/client/recommendations/sidebar/mobile-app/index.jsx]: hey hey @jeherve I think we should still include those list items. Even if they're available in wp-admin, many people probably don't know about them, and they're still good marketing tidbits for the apps I think.

**Comment by robertf4** [comment]
> The plugin had indeed been installed, but the installation had failed. It would be nice to surface any errors in the UI.

@jeffgolenski is currently mocking that up. Once I have the mockup I'll decide if it should be done in this PR or a subsequent one.

**Comment by robertf4** [comment]
> The Progress status for the install buttons seems off a bit

I've updated that so that it has a white color and has a better contrast against the button color.

> I think the app store buttons could do with some extra padding

I've added 32px padding above the app store buttons.

> Maybe we should stay focused on "Dashboard" when in recommendations?

This has been fixed.

**Comment by jeffgolenski** [comment]
Posting the plugin install error state concept here for visibility:

If a plugin can't / won't install, we'll throw an error and provide a "manual install" option. Clicking / tapping that will take the customer directly to the plugin page within the plugin repo



Screen Shot 2021-01-21 at 1 03 14 PM

**Review by jeherve — CHANGES_REQUESTED** [review]
Thanks for the updates! This seems to test well for me at the moment. I noticed 2 things:

Looks like the linting check fails at the moment, for a small thing:
https://github.com/Automattic/jetpack/pull/18437/checks?check_run_id=1743317466

That's going to block the merge; do you think you could fix that?

*********

I noticed that the price of the backup product that was offered to me was different from the price I got when I clicked through:

image
image

It would be nice to fix that.

**Comment by jeffgolenski** [comment]
**Design Feedback w/ Video!**

Thanks for your hard work @robertf4. It's looking outstanding, I've just a few small details to clean up. I've documented each in video below:

**Large screens:**
video -> https://cloudup.com/cdnYwyeMrG7

items:
00:12 - option name php errors
00:28 - need to suppress any JITM messages on the "recommendations" page
00:40 - show recommendation screen on wp-admin dashboard
1:00 - increase description font size to 16px
1:55 - add more padding between right side of content area and image
3:08 - vertical alignment between star icon and "Recommended premium product" headline
3:25 - Recommended premium product card: add a couple px of space between "learn more" and new window icon
3:35 - make the "i" icon a lighter gray. "Gray 40" `#787C82`
3:44 - less space between "View all jetpack features" and new window icon. Let's make all of these consistent throughout entire experience for all "learn more" links on every screen.
3:54 - vertical gray border / divider is darker than the rest
4:04 - small white gap on the right side of the page. -1px right margin cleans it up, but there's probably a more semantic way to fix!
5:00 - "learn more" should open in new tab / window (I know I hate that idea too, but we don't have a clear flow to bring them back from jetpack.com!)
**5:00—5:30 - WooCommerce disappeared from the enabled list when I navigated out of wp-admin and back!**
6:15 - Namining inconsistency (my fault). Monitor is called "Downtime monitoring" in settings. We should make it all consistent.



**Mobile:** 

video -> https://cloudup.com/chfGAOTB6xm

items:
00:20 - padding inconsistency with top and bottom padding of checkboxes & labels
00:35 - bump up size of desc text as mentioned above!
1:00 - line height inconsistency in description text when "learn more" is present
1:35 - alignment issue of settings buttons on final page
1:50 - Originally recommend by @keoshi in a design review, I think we should make the feature titles become the "learn more" link, and actually remove the "learn more" links in our lists to free up space. See video for details
2:19 - reduce padding around "view all jetpack features" box to align with everything else around it

Thanks!

**Comment by robertf4** [comment]
> I noticed that the price of the backup product that was offered to me was different from the price I got when I clicked through

@jeherve I think this is due to a mismatch between the locale being used on the Jetpack site and on the wpcom user account. 

The call to retrieve the upsell price sends the locale from the Jetpack site [here](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php#L824), so the price displayed in the plugin should be based on the locale of the site. The clickthrough goes to cloud.jetpack.com, and there it will use the locale of the wpcom user because that is Calypso.

So I see 3 options here:
1. When retrieving the upsell price use the locale of the wpcom user account. This could result in prices/copy in the upsell card that do not match the language of the rest of the Jetpack plugin.
2. Use the Jetpack site's locale on cloud.jetpack.com. This could result in inconsistency if the user returns to cloud.jetpack.com through a different flow.
3. Leave it as-is, where there will be inconsistency if the Jetpack site and wpcom account have different locales set.

All of these are doable, but I'm not sure which one will create consistency for the greatest number of users. What are your thoughts?

**Comment by jeherve** [comment]
> When retrieving the upsell price use the locale of the wpcom user account. This could result in prices/copy in the upsell card that do not match the language of the rest of the Jetpack plugin.

I would personally vote for this solution, but I'll be happy to let others weigh in with their decision on this.

**Comment by robertf4** [comment]
@jeffgolenski thank you for the review! I was able to address each point except for these ones.

> 00:28 - need to suppress any JITM messages on the "recommendations" page

Suppression of JITMs isn't done in the client, they should just not be configured to show on the recommendations page. I'm not sure how that one is showing in your video because they show based on the URL, and since the recommendations page has its own URL and is new there should be no JITMs for it. For example that particular JITM is set to only show at the root, dashboard, and media URLs. I haven't been able to get it to show on the recommendations page.

> 00:40 - show recommendation screen on wp-admin dashboard

This is coming in this PR: https://github.com/Automattic/jetpack/pull/18359

> WooCommerce disappeared from the enabled list when I navigated out of wp-admin and back!

I am not able to reproduce this. My only thought is that the plugins need a REST call display, and since jurassic ninja sites give 503 errors when anybody spins up a new site it is possible that you loaded it at just the right time for the call to the plugins endpoint to give a 503. Let me know if you are able to reproduce this.

> 1:00 - line height inconsistency in description text when "learn more" is present

The line height is set to 1.5 for everything here so I am not sure how to resolve this. Could you share the CSS that would resolve this?

**Comment by robertf4** [comment]
With regard to the plugin install error states I would like to move that to another PR as I think the recommendations can ship without it if needed, and I think finishing the banner is a higher priority.

**Comment by jeffgolenski** [comment]
@robertf4 In regards to the line-height stuff, we can try to give the `a` links `display: inline-block;` which may clear them up.

**Review by jeherve — APPROVED** [review]
This tests well for me. I only seem to have troubles with the recommendation status.

> Jetpack plugin upgrade flow:
> [...]
> Verify that the Recommendations are not available from the dashboard, and that #/recommendations redirects to #/dashboard

This doesn't seem to work for me; the recommandations tab is always available to me:

- Even upgrading from a site where I had completed the wizard
- Even when the site is in Offline mode #18527  
- When navigating away from the React dashboard and then coming back to it, the recommendations are at the starting step again.

I think this can be solved in a follow-up PR though.

**Review by jeherve — CHANGES_REQUESTED** [review]
I see that we have some linting errors that will need to be looked at before we can merge:
https://github.com/Automattic/jetpack/pull/18437/checks?check_run_id=1752060189
https://github.com/Automattic/jetpack/pull/18437/checks?check_run_id=1752061977

**Review by jeherve — APPROVED** [review]
Merging this now, so we can include it in 9.4.

**Comment by jeherve** [comment]
> This doesn't seem to work for me; the recommandations tab is always available to me:
>
> Even upgrading from a site where I had completed the wizard
> ~~Even when the site is in Offline mode #18527~~
> When navigating away from the React dashboard and then coming back to it, the recommendations are at the starting step again.

Would you like me to create a follow-up issue to keep track of this, or is this already on your radar?

**Comment by jeherve** [comment]
r220018-wpcom

**Pull request merged**
